### PR TITLE
Speed improvement

### PR DESCRIPTION
**Pull request by jannisborn**
Starting to work on speed improvements for the StackGRU-VAE:

Here are the logs for current speed tests:

```
INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.003, Encoder:0.313, Decoder: 0.172

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 500
Setup: 0.003, Encoder:0.333, Decoder: 0.199

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 256, Vocab: 100
Setup: 0.016, Encoder:0.475, Decoder: 0.243

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 256, Vocab: 500
Setup: 0.020, Encoder:0.494, Decoder: 0.291

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 128, RNN: 16, Vocab: 100
Setup: 0.002, Encoder:2.407, Decoder: 1.351

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 128, RNN: 16, Vocab: 500
Setup: 0.006, Encoder:4.654, Decoder: 3.753

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 128, RNN: 256, Vocab: 100
Setup: 0.028, Encoder:6.928, Decoder: 3.738

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 128, RNN: 256, Vocab: 500
Setup: 0.051, Encoder:5.802, Decoder: 3.201

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 4, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.014, Encoder:1.327, Decoder: 0.563

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 4, batch_size: 8, RNN: 16, Vocab: 500
Setup: 0.004, Encoder:0.802, Decoder: 0.345

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 4, batch_size: 8, RNN: 256, Vocab: 100
Setup: 0.036, Encoder:0.915, Decoder: 0.461

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 4, batch_size: 8, RNN: 256, Vocab: 500
Setup: 0.043, Encoder:1.096, Decoder: 0.509

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 4, batch_size: 128, RNN: 16, Vocab: 100
Setup: 0.004, Encoder:1.957, Decoder: 3.169

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 4, batch_size: 128, RNN: 16, Vocab: 500
Setup: 0.011, Encoder:3.633, Decoder: 1.861

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 4, batch_size: 128, RNN: 256, Vocab: 100
Setup: 0.044, Encoder:7.843, Decoder: 3.790

INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 4, batch_size: 128, RNN: 256, Vocab: 500
Setup: 0.054, Encoder:6.976, Decoder: 4.456

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.007, Encoder:1.130, Decoder: 0.587

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 500
Setup: 0.007, Encoder:0.435, Decoder: 0.263

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 256, Vocab: 100
Setup: 0.021, Encoder:0.679, Decoder: 0.279

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 256, Vocab: 500
Setup: 0.022, Encoder:0.602, Decoder: 0.278

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 2, batch_size: 128, RNN: 16, Vocab: 100
Setup: 0.003, Encoder:1.421, Decoder: 0.793

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 2, batch_size: 128, RNN: 16, Vocab: 500
Setup: 0.003, Encoder:1.522, Decoder: 1.229

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 2, batch_size: 128, RNN: 256, Vocab: 100
Setup: 0.017, Encoder:3.483, Decoder: 3.746

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 2, batch_size: 128, RNN: 256, Vocab: 500
Setup: 0.071, Encoder:7.022, Decoder: 2.598

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 4, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.012, Encoder:0.967, Decoder: 0.491

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 4, batch_size: 8, RNN: 16, Vocab: 500
Setup: 0.003, Encoder:0.623, Decoder: 0.340

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 4, batch_size: 8, RNN: 256, Vocab: 100
Setup: 0.033, Encoder:0.957, Decoder: 0.462

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 4, batch_size: 8, RNN: 256, Vocab: 500
Setup: 0.045, Encoder:2.901, Decoder: 0.706

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 4, batch_size: 128, RNN: 16, Vocab: 100
Setup: 0.007, Encoder:6.007, Decoder: 2.243

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 4, batch_size: 128, RNN: 16, Vocab: 500
Setup: 0.007, Encoder:3.173, Decoder: 2.199

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 4, batch_size: 128, RNN: 256, Vocab: 100
Setup: 0.042, Encoder:8.101, Decoder: 4.945

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 4, batch_size: 128, RNN: 256, Vocab: 500
Setup: 0.157, Encoder:9.307, Decoder: 4.686

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.007, Encoder:0.926, Decoder: 0.456

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 500
Setup: 0.005, Encoder:0.190, Decoder: 0.145

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 256, Vocab: 100
Setup: 0.016, Encoder:0.418, Decoder: 0.235

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 256, Vocab: 500
Setup: 0.028, Encoder:0.353, Decoder: 0.238

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 2, batch_size: 128, RNN: 16, Vocab: 100
Setup: 0.002, Encoder:0.330, Decoder: 0.255

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 2, batch_size: 128, RNN: 16, Vocab: 500
Setup: 0.003, Encoder:0.310, Decoder: 0.655

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 2, batch_size: 128, RNN: 256, Vocab: 100
Setup: 0.017, Encoder:1.517, Decoder: 0.897

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 2, batch_size: 128, RNN: 256, Vocab: 500
Setup: 0.021, Encoder:1.571, Decoder: 1.234

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 4, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.003, Encoder:0.376, Decoder: 0.234

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 4, batch_size: 8, RNN: 16, Vocab: 500
Setup: 0.003, Encoder:0.344, Decoder: 0.203

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 4, batch_size: 8, RNN: 256, Vocab: 100
Setup: 0.034, Encoder:0.646, Decoder: 0.464

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 4, batch_size: 8, RNN: 256, Vocab: 500
Setup: 0.049, Encoder:0.657, Decoder: 0.346

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 4, batch_size: 128, RNN: 16, Vocab: 100
Setup: 0.004, Encoder:0.444, Decoder: 0.334

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 4, batch_size: 128, RNN: 16, Vocab: 500
Setup: 0.004, Encoder:0.472, Decoder: 0.731

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 4, batch_size: 128, RNN: 256, Vocab: 100
Setup: 0.034, Encoder:2.692, Decoder: 1.796

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 4, batch_size: 128, RNN: 256, Vocab: 500
Setup: 0.042, Encoder:8.421, Decoder: 2.973

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.007, Encoder:0.528, Decoder: 0.289

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 500
Setup: 0.005, Encoder:0.220, Decoder: 0.156

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 256, Vocab: 100
Setup: 0.019, Encoder:0.394, Decoder: 0.203

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 256, Vocab: 500
Setup: 0.026, Encoder:0.341, Decoder: 0.183

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 2, batch_size: 128, RNN: 16, Vocab: 100
Setup: 0.003, Encoder:0.280, Decoder: 0.252

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 2, batch_size: 128, RNN: 16, Vocab: 500
Setup: 0.003, Encoder:0.276, Decoder: 0.631

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 2, batch_size: 128, RNN: 256, Vocab: 100
Setup: 0.017, Encoder:1.297, Decoder: 0.846

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 2, batch_size: 128, RNN: 256, Vocab: 500
Setup: 0.025, Encoder:1.750, Decoder: 1.361

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 4, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.003, Encoder:0.403, Decoder: 0.223

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 4, batch_size: 8, RNN: 16, Vocab: 500
Setup: 0.003, Encoder:0.322, Decoder: 0.190

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 4, batch_size: 8, RNN: 256, Vocab: 100
Setup: 0.032, Encoder:0.490, Decoder: 0.263

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 4, batch_size: 8, RNN: 256, Vocab: 500
Setup: 0.040, Encoder:0.562, Decoder: 0.301

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 4, batch_size: 128, RNN: 16, Vocab: 100
Setup: 0.003, Encoder:0.470, Decoder: 0.329

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 4, batch_size: 128, RNN: 16, Vocab: 500
Setup: 0.004, Encoder:0.491, Decoder: 0.738

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 4, batch_size: 128, RNN: 256, Vocab: 100
Setup: 0.034, Encoder:2.683, Decoder: 1.846

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 4, batch_size: 128, RNN: 256, Vocab: 500
Setup: 0.043, Encoder:4.554, Decoder: 2.368

INFO:test_vae:***Decoder search tests***
INFO:test_vae:Search: GreedySearch, Generated length: 50,        (BS:8, #Layer: 2, RNN 16)
 Took: 0.068
INFO:test_vae:Search: SamplingSearch, Generated length: 50,      (BS:8, #Layer: 2, RNN 16)
 Took: 0.068
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 2, Top-K: 5   (BS:8, #Layer: 2, RNN 16)
 Took: 0.300
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 2, Top-K: 30          (BS:8, #Layer: 2, RNN 16)
 Took: 0.608
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 8, Top-K: 5   (BS:8, #Layer: 2, RNN 16)
 Took: 0.623
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 8, Top-K: 30          (BS:8, #Layer: 2, RNN 16)
 Took: 2.297
INFO:test_vae:Search: GreedySearch, Generated length: 200,       (BS:8, #Layer: 2, RNN 16)
 Took: 0.191
INFO:test_vae:Search: SamplingSearch, Generated length: 200,     (BS:8, #Layer: 2, RNN 16)
 Took: 0.236
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 2, Top-K: 5          (BS:8, #Layer: 2, RNN 16)
 Took: 1.711
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 2, Top-K: 30         (BS:8, #Layer: 2, RNN 16)
 Took: 3.124
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 8, Top-K: 5          (BS:8, #Layer: 2, RNN 16)
 Took: 2.610
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 8, Top-K: 30         (BS:8, #Layer: 2, RNN 16)
 Took: 8.486
INFO:test_vae:Search: GreedySearch, Generated length: 50,        (BS:8, #Layer: 2, RNN 16)
 Took: 0.049
INFO:test_vae:Search: SamplingSearch, Generated length: 50,      (BS:8, #Layer: 2, RNN 16)
 Took: 0.054
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 2, Top-K: 5   (BS:8, #Layer: 2, RNN 16)
 Took: 0.243
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 2, Top-K: 30          (BS:8, #Layer: 2, RNN 16)
 Took: 0.546
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 8, Top-K: 5   (BS:8, #Layer: 2, RNN 16)
 Took: 0.593
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 8, Top-K: 30          (BS:8, #Layer: 2, RNN 16)
 Took: 2.054
INFO:test_vae:Search: GreedySearch, Generated length: 200,       (BS:8, #Layer: 2, RNN 16)
 Took: 0.179
INFO:test_vae:Search: SamplingSearch, Generated length: 200,     (BS:8, #Layer: 2, RNN 16)
 Took: 0.207
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 2, Top-K: 5          (BS:8, #Layer: 2, RNN 16)
 Took: 0.952
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 2, Top-K: 30         (BS:8, #Layer: 2, RNN 16)
 Took: 2.197
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 8, Top-K: 5          (BS:8, #Layer: 2, RNN 16)
 Took: 2.491
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 8, Top-K: 30         (BS:8, #Layer: 2, RNN 16)
 Took: 11.493
INFO:test_vae:Search: GreedySearch, Generated length: 50,        (BS:8, #Layer: 2, RNN 16)
 Took: 0.023
INFO:test_vae:Search: SamplingSearch, Generated length: 50,      (BS:8, #Layer: 2, RNN 16)
 Took: 0.032
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 2, Top-K: 5   (BS:8, #Layer: 2, RNN 16)
 Took: 0.205
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 2, Top-K: 30          (BS:8, #Layer: 2, RNN 16)
 Took: 0.522
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 8, Top-K: 5   (BS:8, #Layer: 2, RNN 16)
 Took: 0.463
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 8, Top-K: 30          (BS:8, #Layer: 2, RNN 16)
 Took: 1.876
INFO:test_vae:Search: GreedySearch, Generated length: 200,       (BS:8, #Layer: 2, RNN 16)
 Took: 0.092
INFO:test_vae:Search: SamplingSearch, Generated length: 200,     (BS:8, #Layer: 2, RNN 16)
 Took: 0.112
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 2, Top-K: 5          (BS:8, #Layer: 2, RNN 16)
 Took: 0.745
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 2, Top-K: 30         (BS:8, #Layer: 2, RNN 16)
 Took: 2.564
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 8, Top-K: 5          (BS:8, #Layer: 2, RNN 16)
 Took: 1.780
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 8, Top-K: 30         (BS:8, #Layer: 2, RNN 16)
 Took: 8.610
INFO:test_vae:Search: GreedySearch, Generated length: 50,        (BS:8, #Layer: 2, RNN 16)
 Took: 0.033
INFO:test_vae:Search: SamplingSearch, Generated length: 50,      (BS:8, #Layer: 2, RNN 16)
 Took: 0.046
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 2, Top-K: 5   (BS:8, #Layer: 2, RNN 16)
 Took: 0.252
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 2, Top-K: 30          (BS:8, #Layer: 2, RNN 16)
 Took: 0.579
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 8, Top-K: 5   (BS:8, #Layer: 2, RNN 16)
 Took: 0.561
INFO:test_vae:Search: BeamSearch, Generated length: 50, Beam size: 8, Top-K: 30          (BS:8, #Layer: 2, RNN 16)
 Took: 2.275
INFO:test_vae:Search: GreedySearch, Generated length: 200,       (BS:8, #Layer: 2, RNN 16)
 Took: 0.099
INFO:test_vae:Search: SamplingSearch, Generated length: 200,     (BS:8, #Layer: 2, RNN 16)
 Took: 0.122
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 2, Top-K: 5          (BS:8, #Layer: 2, RNN 16)
 Took: 0.863
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 2, Top-K: 30         (BS:8, #Layer: 2, RNN 16)
 Took: 2.041
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 8, Top-K: 5          (BS:8, #Layer: 2, RNN 16)
 Took: 1.891
INFO:test_vae:Search: BeamSearch, Generated length: 200, Beam size: 8, Top-K: 30         (BS:8, #Layer: 2, RNN 16)
 Took: 7.883
```

### CODE REVIEW

**Comment by jannisborn** [comment]
Cleaned version:
```
INFO:test_vae:Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.002, Encoder:0.331, Decoder: 0.166

INFO:test_vae:Stack: True, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.001, Encoder:0.194, Decoder: 0.175

INFO:test_vae:Stack: False, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.002, Encoder:0.164, Decoder: 0.101

INFO:test_vae:Stack: False, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 16, Vocab: 100
Setup: 0.001, Encoder:0.094, Decoder: 0.099
```
Tl;Dr: Bidirectional = True makes the encoder amost 50% slower. The same applies to using stack or not using it. No stack and no bidirectional encoder needs only 40% of the time of using both.

**Comment by jannisborn** [comment]
Overall numbers look better and it is faster now. Individual numbers are not reliable though, they come from my local machine and I didn't repeat/average test speeds.

**PADDED**

```
INFO:test_vae:      Mode Padded, Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 20, Vocab: 100
Setup: 0.002, Encoder:0.216, Decoder: 0.129

INFO:test_vae:  Mode Padded, Stack: True, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 20, Vocab: 100
Setup: 0.002, Encoder:0.494, Decoder: 0.297

INFO:test_vae:  Mode Padded, Stack: False, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 20, Vocab: 100
Setup: 0.003, Encoder:0.310, Decoder: 0.186

INFO:test_vae:  Mode Padded, Stack: False, bidirectional: False SeqLen:200, layers: 2, batch_size: 8, RNN: 20, Vocab: 100
Setup: 0.002, Encoder:0.342, Decoder: 0.209

```


*** PACKED ***

```
INFO:test_vae:  Mode Packed, Stack: True, bidirectional: True SeqLen:200, layers: 2, batch_size: 8, RNN: 32, Vocab: 100
Setup: 0.002, Encoder: 0.172, Reparam: 0.000 Decoder: 0.128

**Comment by jannisborn** [comment]
- @niladell  added support for packed sequences
- refactoring many small things in `StackGRU`, `StackGRUEncoder` and `StackGRUDecoder` for speed enhancement.

@niladell @drugilsberg please review and/or merge.

File: paccmann_chemistry/models/training.py [hunk 1]
```diff
@@ -142,6 +146,11 @@ def train_vae(
         )
         loss.backward()
         train_loss += loss.detach().item()
+        # for name, param in vae_model.named_parameters():
+        #     if param.grad is None:
+        #         print(name)
+        #         print(param.grad)
```
Comment (niladell) [inline paccmann_chemistry/models/training.py]: Should this commented section be removed?

File: paccmann_chemistry/models/vae.py [hunk 2]
```diff
@@ -182,18 +313,80 @@ def decoder_train_step(self, latent_z, input_seq, target_seq):
...
+        if not isinstance(input_seq, nn.utils.rnn.PackedSequence):
+            raise TypeError('Input is not PackedSequence')
+
+        loss = 0
+        input_seq_packed, batch_sizes = utils.perpare_packed_input(input_seq)
+        # Target sequence should have same batch_sizes as input_seq
+        input_seq_packed, _ = utils.perpare_packed_input(target_seq)
+        prev_batch = batch_sizes[0]
+
+        outputs = []
```
Comment (niladell) [inline paccmann_chemistry/models/vae.py]: Do we need to store the outputs?

File: paccmann_chemistry/utils/utils.py [hunk 3]
```diff
@@ -67,6 +68,71 @@ def sequential_data_preparation(
...
+        if input_keep != 1:
+            # mask for token dropout
+            mask = Bernoulli(input_keep).sample(input.shape)
+            mask = torch.LongTensor(mask.numpy())
+            dropout_loc = np.where(mask == 0)[0]
+            decoder[dropout_loc] = dropout_index
+
+        # TODO: Don't we have the start index at the begining already?
+        # input_seq = torch.cat((torch.tensor([start_index]), input))
+        # decoder_seq = torch.cat((torch.tensor([start_index]), decoder))
```
Comment (niladell) [inline paccmann_chemistry/utils/utils.py]: I left this TODO here when I merged, and no I am not sure if what I was doing here is correct. @jannisborn I did remove this part here because the sequences I was passing in already had the start and end index. We should double check that and if the start/end token are already there we can remove them from this fn (or if they are not un-comment this part). In any case, there may be an inconsistency with the `padded` data prep.

**Comment by niladell** [comment]
Most of the comments are just aesthetic/misc. things, only thing to check properly is the last comment. For the rest it looks cools great to me.

**Review by drugilsberg — APPROVED** [review]
Agree with the comments from @niladell

File: paccmann_chemistry/models/training.py (hunk 1 above)
Comment (jannisborn) [reply paccmann_chemistry/models/training.py]: thx, removed it.

File: paccmann_chemistry/models/vae.py (hunk 2 above)
Comment (jannisborn) [reply paccmann_chemistry/models/vae.py]: nope, only in the `_forward_pass_padded` so that we can look at an example reconstruction (for `_forward_pass_packed` that's a bit trickier, so I omit it for the moment)

File: paccmann_chemistry/utils/utils.py (hunk 3 above)
Comment (jannisborn) [reply paccmann_chemistry/utils/utils.py]: you're right. This should be controlled only in the `SMILESDataset` and the `SMILESLanguage`. I removed the addition of the start_index from both data preparation methods and also refactored/speeded it up a bit.
